src/main/java/com/tterrag/registrate/AbstractRegistrate.java

```diff
@@ -123,6 +124,79 @@ void addRegisterCallback(NonNullConsumer<? super T> callback) {
+        private final Supplier<FeatureFlagSet> flags;
+        private final BooleanSupplier hasPermissions;
+        private final BiConsumer<ItemStack, CreativeModeTab.TabVisibility> acceptFunc;
+
+        private CreativeModeTabModifier(Supplier<FeatureFlagSet> flags, BooleanSupplier hasPermissions, BiConsumer<ItemStack, CreativeModeTab.TabVisibility> acceptFunc) {
+            this.flags = flags;
+            this.hasPermissions = hasPermissions;
+            this.acceptFunc = acceptFunc;
+        }
+
+        public FeatureFlagSet getFlags() {
+            return flags.get();
+        }
+
+        public boolean hasPermissions() {
+            return hasPermissions.getAsBoolean();
+        }
+
+        @Override
+        public void accept(ItemStack stack, CreativeModeTab.TabVisibility visibility) {
+            acceptFunc.accept(stack, visibility);
+        }
+
+        public void accept(Supplier<? extends ItemLike> item, CreativeModeTab.TabVisibility visibility) {
+            accept(item.get(), visibility);
+        }
+
+        public void accept(Supplier<? extends ItemLike> item) {
+            accept(item.get());
+        }
+    }
+
+    private static final class CreativeModeTabRegistration implements Supplier<CreativeModeTab> {
+        private static final Supplier<List<Object>> DEFAULT_AFTER_ENTRIES = Lazy.of(CreativeModeTabRegistration::defaultAfterEntries);
+
+        @Nullable private CreativeModeTab creativeModeTab = null;
+        private final ResourceLocation registryName;
+        private final List<Object> beforeEntries;
+        private final List<Object> afterEntries;
+        private final Consumer<CreativeModeTab.Builder> configurator;
+
+        private CreativeModeTabRegistration(ResourceLocation registryName, @Nullable List<Object> beforeEntries, @Nullable List<Object> afterEntries, Consumer<CreativeModeTab.Builder> configurator) {
+            this.registryName = registryName;
+            this.beforeEntries = beforeEntries == null ? List.of() : beforeEntries;
+            this.afterEntries = afterEntries == null ? DEFAULT_AFTER_ENTRIES.get() : afterEntries;
+            this.configurator = configurator;
+        }
+
+        void register(CreativeModeTabEvent.Register event) {
+            log.debug("Registering CreativeModeTab '{}'", registryName);
+            creativeModeTab = event.registerCreativeModeTab(registryName, beforeEntries, afterEntries, configurator);
+        }
+
+        @Override
+        public CreativeModeTab get() {
+            return Objects.requireNonNull(creativeModeTab, () -> "Attempt to obtain CreativeModeTab(%s) instance before it was registered!".formatted(registryName));
+        }
+
+        @SuppressWarnings("unchecked")
+        private static List<Object> defaultAfterEntries() {
+            try {
+                var field = CreativeModeTabEvent.class.getDeclaredField("DEFAULT_AFTER_ENTRIES");
```

Comment:
It is simpler to use `ObfuscationReflectionHelper` here, even if the target is not obfuscated.